### ISSUE TITLE
Inheriting Validation decorators

### ISSUE THREAD

**Issue by shymanel** [typestack/class-validator#633]
### Description

From the [docs](https://github.com/typestack/class-validator#inheriting-validation-decorators):

If a property is redefined in the descendant class decorators will be applied on it **both from that and the base class**.

I'm not sure it works as it should.

### Reproduction

```ts
import { MinLength, IsUppercase } from "class-validator";
import { validate } from "class-validator";

class BaseClass {
    @MinLength(10)
    someProperty: string;
}

class ChildClass extends BaseClass {
    @IsUppercase()
    someProperty: string;
}

const someObj = new ChildClass();
someObj.someProperty = 'HELLO';

validate(someObj).then(result => {
    console.log(result);
});

// the output is empty array []
```

From what I'm reading in docs, I thought `@MinLength(10)`  is used alongside `@IsUppercase()`, but I don't see it. Moreover, I'm not sure what empty array means.

### Environment

- [ ] nodejs: v12.17.0
- [ ] typescript: 3.9.5

**Comment by cduff** [comment]
I'm seeing the same. It does not appear to behave as documented.

@shymanel The empty array output from your example indicates no validation errors.

**Comment by shymanel** [comment]
@cduff yep, thanks, I figured it out. I though there should be an error, since `HELLO` is not 10 characters long

**Comment by cduff** [comment]
FYI, since commenting on this issue earlier I've been using class-validator more and have come to realise that not having it inherit decorators is useful sometimes.

For example, I can have code like:

```
class User {
  @Matches(hashPattern)
  passwordHash?: string;
}

class RegistrationRequest extends User {
  @Length(minLength, maxLength)
  password?: string;

  @IsEmpty()
  passwordHash?: string;
}
```

For a `User` a valid `passwordHash` is required. But for a `RegistrationRequest` a valid `password` should be supplied instead, from which the server will generate a valid `passwordHash`. This wouldn't work if the `Matches` decorator was inherited to the `RegistrationRequest.passwordHash` property. To achieve the same would require some verbose `groups` usage I imagine, noting that in reality the `User` has many other properties not shown above.

So now I'm wondering if the decorators are not inheriting by design?

**Comment by ragrub** [comment]
I encounter the same issue described by @shymanel. 
At least with version 0.11.1 it used to work as mentioned in the documentation, that decorators were inherited from base class.

https://github.com/typestack/class-validator#inheriting-validation-decorators 
"user.password = 'too short'; // password wil be validated not only against IsString, but against MinLength as well"

I would like to update to version 0.12.2, but this seems to be broken now and affects a lot of unit tests in my project.

**Comment by NickKelly1** [comment]
There is now a [PR](https://github.com/typestack/class-validator/pull/748) for this

**Comment by cduff** [comment]
Any update on this? I got hit by it again today. It's quite confusing at the moment as some decorators are inherited (e.g. @IsOptional) and others aren't.

**Comment by Arcdie** [comment]
While we are waiting for merge, you can make your own decorator and call it in child class.

```
import {getMetadataStorage} from 'class-validator';

function inheritParentDecorators() {
  return (
    target: any,
    propertyKey: string,
  ) => {
    const storage = getMetadataStorage();
    const parent = Object.getPrototypeOf(target.constructor);

    if (!parent) {
      return;
    }

    const targetMetadatas = storage.getTargetValidationMetadatas(
      parent,
      parent.name,
      false,
      false,
    );

    targetMetadatas
      .filter(e => e.propertyName === propertyKey)
      .forEach(e => {
        registerDecorator({
          name: e.type,
          target: target.constructor,
          propertyName: e.propertyName,
          // preserving custom error messages (thanks [slavco86](https://github.com/slavco86))
          options: {...e.validationTypeOptions, message: e.message},
          constraints: e.constraints,
          validator: e.constraintCls,
        });
      });
  };
}

class BaseClass {
    @MinLength(10)
    someProperty: string;
}

class ChildClass extends BaseClass {
    @inheritParentDecorators()
    @IsUppercase()
    someProperty: string;
}
```

**Comment by slavco86** [comment]
Please, please, please - someone pick this up urgently. This is a clear regression from V11 and we are on V13 already without this being fixed.
🙏

**Comment by slavco86** [comment]
> While we are waiting for merge, you can make your own decorator and call it in child class.
> 
> ```
> import {getMetadataStorage} from 'class-validator';
> 
> function inheritParentDecorators() {
>   return (
>     target: any,
>     propertyKey: string,
>   ) => {
>     const storage = getMetadataStorage();
>     const parent = Object.getPrototypeOf(target.constructor);
> 
>     if (!parent) {
>       return;
>     }
> 
>     const targetMetadatas = storage.getTargetValidationMetadatas(
>       parent,
>       parent.name,
>       false,
>       false,
>     );
> 
>     targetMetadatas
>       .filter(e => e.propertyName === propertyKey)
>       .forEach(e => {
>         registerDecorator({
>           name: e.type,
>           target: target.constructor,
>           propertyName: e.propertyName,
>           options: e.validationTypeOptions,
>           constraints: e.constraints,
>           validator: e.constraintCls,
>         });
>       });
>   };
> }
> 
> class BaseClass {
>     @MinLength(10)
>     someProperty: string;
> }
> 
> class ChildClass extends BaseClass {
>     @inheritParentDecorators()
>     @IsUppercase()
>     someProperty: string;
> }
> ```

Slight, update to preserve any custom error messages, potentially defined in the parent decorator options:
```
import {getMetadataStorage} from 'class-validator';

function inheritParentDecorators() {
  return (
    target: any,
    propertyKey: string,
  ) => {
    const storage = getMetadataStorage();
    const parent = Object.getPrototypeOf(target.constructor);

    if (!parent) {
      return;
    }

    const targetMetadatas = storage.getTargetValidationMetadatas(
      parent,
      parent.name,
      false,
      false,
    );

    targetMetadatas
      .filter(e => e.propertyName === propertyKey)
      .forEach(e => {
        registerDecorator({
          name: e.type,
          target: target.constructor,
          propertyName: e.propertyName,
          options: {...e.validationTypeOptions, message: e.message},
          constraints: e.constraints,
          validator: e.constraintCls,
        });
      });
  };
}

class BaseClass {
    @MinLength(10)
    someProperty: string;
}

class ChildClass extends BaseClass {
    @inheritParentDecorators()
    @IsUppercase()
    someProperty: string;
}
```

@Arcdie , if you want to include `options: {...e.validationTypeOptions, message: e.message},` as part of your response - I will remove my comment to make it clearer and more straightforward for everyone who will want (like I did) to use your brilliant work around in the mean time

**Comment by Arcdie** [comment]
@[slavco86](https://github.com/slavco86), edited, thanks for tip.

**Comment by arturohernandez10** [comment]
Just FYI.

It only makes sense to inherit if the decorators don't conflict with each other.

**Comment by huybuidac** [comment]
> PR

hey bro, how did you fix this :(

**Comment by josephdpurcell** [comment]
The documentation says inheritance should work: https://github.com/typestack/class-validator/issues/633. But, it doesn't for me, I get:

> Parameters violate constraints: policy must be one of the following values: [object Object]

I tried the solution from https://github.com/typestack/class-validator/issues/633#issuecomment-1110790596 and got the same error.

This is from `@IsIn` decorator on the parent class, but same error happens with other validators. I haven't dug in to find out why.

Within NestJS I've found a solution/hack that does the inheritance:

```ts
import { OmitType } from "@nestjs/swagger";

class ChildClass extends OmitType(BaseClass, []) {
    @IsUppercase()
    someProperty: string;
}
```

For now I can use the type helpers (pick/omit) from `@nestjs/swagger` but it would be great to see a better way to do this, even if its a custom decorator like `@inheritParentDecorators`.

Edit: clarified that I've read the docs.

**Comment by elbasan** [comment]
this is still an issue, how come nobody from the dev team has looked into it?

**Comment by cduff** [comment]
Great to see some activity and a [new release](https://github.com/typestack/class-validator/releases/tag/v0.14.2) for class-validator recently!

Any chance of getting this issue resolved? @braaar

**Comment by cduff** [comment]
An interesting discussion around this issue going on [here](https://github.com/typestack/class-validator/discussions/2673).

Looks like others are using a fork of `class-validator` to work around this issue. Example:

https://github.com/proconnect-gouv/federation/pull/696

With a new version of `class-validator` being released as recently as [last month](https://github.com/typestack/class-validator/releases/tag/v0.15.1), it'd be great to finally get this major issue resolved!

@braaar It would be great to get your thoughts on whether this might be a possibility, or whether we should be looking at forks or alternative solutions?

Thanks @braaar for your work on `class-validator`!
